src/category_theory/idempotents/functor_categories.lean

```diff
@@ -0,0 +1,167 @@
+			(by rw [comp_id, assoc, p.naturality φ, ← assoc, ← limits.equalizer.condition, comp_id]),
+		map_id' := λ j, by { ext, simp only [comp_id, functor.map_id, equalizer.lift_ι, id_comp], },
+		map_comp' := λ j j' j'' φ φ', begin
+			ext,
+			simp only [assoc, functor.map_comp, equalizer.lift_ι, equalizer.lift_ι_assoc],
+		end },
+	let i : Y ⟶ F :=
+	{ app := λ j, equalizer.ι _ _,
+		naturality' := λ j j' φ, by rw [equalizer.lift_ι],	},
+	let e : F ⟶ Y :=
+	{ app := λ j, equalizer.lift (p.app j)
+			(by { rw comp_id, exact (congr_app hp j).symm, }),
+		naturality' := λ j j' φ, begin
+			ext,
+			simp only [assoc, equalizer.lift_ι, nat_trans.naturality, equalizer.lift_ι_assoc],
+		end },
+	use [Y, i, e],
+	split; ext j,
+	{ simp only [nat_trans.comp_app, assoc, equalizer.lift_ι, nat_trans.id_app, id_comp,
+			← equalizer.condition, comp_id], },
+	{ simp only [nat_trans.comp_app, equalizer.lift_ι], },
+end
+
+namespace karoubi_functor_category_embedding
+
+variables {J C}
+
+/-- On objects, the functor which sends a formal direct factor `P` of a
+functor `F : J ⥤ C` to the functor `J ⥤ karoubi C` which sends `(j : J)` to
+the corresponding direct factor of `F.obj j`. -/
+@[simps]
+def obj (P : karoubi (J ⥤ C)) : J ⥤ karoubi C :=
+{ obj := λ j, ⟨P.X.obj j, P.p.app j, congr_app P.idempotence j⟩,
+	map := λ j j' φ,
+	{ f := P.p.app j ≫ P.X.map φ,
+		comm := begin
+			simp only [nat_trans.naturality, assoc],
+     	have h := congr_app P.idempotence j,
+     	rw [nat_trans.comp_app] at h,
+     	slice_rhs 1 3 { erw [h, h], },
+		end },
+	map_id' := λ j, by { ext, simp only [functor.map_id, comp_id, id_eq], },
+	map_comp' := λ j j' j'' φ φ', begin
+		ext,
+   	have h := congr_app P.idempotence j,
+   	rw [nat_trans.comp_app] at h,
+		simp only [assoc, nat_trans.naturality_assoc, functor.map_comp,
+     	comp],
+    slice_rhs 1 2 { rw h, },
+ 	  rw [assoc],
+	end  }
+
+/-- Action of maps on `obj`. -/
+@[simps]
+def map {P Q : karoubi (J ⥤ C)} (f : P ⟶ Q) : obj P ⟶ obj Q :=
+{ app := λ j, ⟨f.f.app j, congr_app f.comm j⟩,
+	naturality' := λ j j' φ, begin
+		ext,
+		simp only [comp],
+    have h := congr_app (comp_p f) j,
+    have h' := congr_app (p_comp f) j',
+ 	  dsimp at h h' ⊢,
+   	slice_rhs 1 2 { erw h, },
+   	rw ← P.p.naturality,
+   	slice_lhs 2 3 { erw h', },
+   	rw f.f.naturality,
+	end }
+
+end karoubi_functor_category_embedding
+
+variables (J C)
+
+/-- The tautological fully faithful functor `karoubi (J ⥤ C) ⥤ (J ⥤ karoubi C)`. -/
+@[simps]
+def karoubi_functor_category_embedding :
+	karoubi (J ⥤ C) ⥤ (J ⥤ karoubi C) :=
+{ obj := karoubi_functor_category_embedding.obj,
+	map := λ P Q, karoubi_functor_category_embedding.map,
+	map_id' := λ P, rfl,
+	map_comp' := λ P Q R f g, rfl, }
+
+instance : full (karoubi_functor_category_embedding J C) :=
+{ preimage := λ P Q f,
+  { f :=
+    { app := λ j, (f.app j).f,
+      naturality' := λ j j' φ, begin
+				slice_rhs 1 1 { rw ← karoubi.comp_p, },
```

Comment:
Remove excess indenting (here and elsewhere).

Reply:
Thanks. I do not know what happened (it looked fine in VSCode, but some spaces were replaced by the Tab character). I have fixed this.